crates/volta-migrate/src/v2.rs

```diff
@@ -0,0 +1,158 @@
+///
+/// Holds a reference to the V1 layout struct to support potential future migrations
+pub struct V2 {
+    pub home: v2::VoltaHome,
+}
+
+impl V2 {
+    pub fn new(home: PathBuf) -> Self {
+        V2 {
+            home: v2::VoltaHome::new(home),
+        }
+    }
+
+    /// Write the layout file to mark migration to V1 as complete
+    ///
+    /// Should only be called once all other migration steps are finished, so that we don't
+    /// accidentally mark an incomplete migration as completed
+    fn complete_migration(home: v2::VoltaHome) -> Fallible<Self> {
+        debug!("Writing layout marker file");
+        File::create(home.layout_file()).with_context(|_| ErrorDetails::CreateLayoutFileError {
+            file: home.layout_file().to_owned(),
+        })?;
+
+        Ok(V2 { home })
+    }
+}
+
+impl TryFrom<Empty> for V2 {
+    type Error = VoltaError;
+
+    fn try_from(old: Empty) -> Fallible<V2> {
+        debug!("New Volta installation detected, creating fresh layout");
+
+        let home = v2::VoltaHome::new(old.home);
+        home.create()
+            .with_context(|_| ErrorDetails::CreateDirError {
+                dir: home.root().to_owned(),
+            })?;
+
+        V2::complete_migration(home)
+    }
+}
+
+impl TryFrom<V1> for V2 {
+    type Error = VoltaError;
+
+    fn try_from(old: V1) -> Fallible<V2> {
+        debug!("Migrating from V1 layout");
+
+        let new_home = v2::VoltaHome::new(old.home.root().to_owned());
+        new_home
+            .create()
+            .with_context(|_| ErrorDetails::CreateDirError {
+                dir: new_home.root().to_owned(),
+            })?;
+
+        // Check the default platform file `platform.json`
+        // If it contains an npm version that matches the default, update it to have None instead
+        // This will ensure that we don't treat the default npm from a prior version of Volta
+        // as a "custom" npm that the user explicitly requested
+        let platform_file = old.home.default_platform_file();
+        if platform_file.exists() {
+            let platform_json = read_to_string(platform_file).with_context(|_| {
+                ErrorDetails::ReadPlatformError {
+                    file: platform_file.to_owned(),
+                }
+            })?;
+            let mut existing_platform = Platform::from_json(platform_json)?;
+
+            if let Some(ref mut node_version) = &mut existing_platform.node {
+                if let Some(npm) = &node_version.npm {
+                    if *npm == load_default_npm_version(&node_version.runtime)? {
+                        node_version.npm = None;
+                        write(platform_file, existing_platform.into_json()?).with_context(
+                            |_| ErrorDetails::WritePlatformError {
+                                file: platform_file.to_owned(),
+                            },
+                        )?;
+                    }
+                }
+            }
+        }
```

Comment:
Similar to my comment below, I wonder if it makes sense to extract this as its own function. I tend to instinctively aim to keep my functions shorter—not quite to the extent of [Sandi Metz's "5-line-methods" rule](https://thoughtbot.com/blog/sandi-metz-rules-for-developers#five-lines-per-method), but _inspired_ by that at a minimum.

Reply:
Agreed, will do. Thanks!